spec/haskell/src/SEL4/Object/SchedContext.lhs

```diff
@@ -345,15 +345,46 @@ This module uses the C preprocessor to select a target architecture.
 > refillHeadOverlappingLoop scPtr =
 >        whileLoop (const (fromJust . runReaderT (refillHeadOverlapping scPtr))) (const (mergeRefills scPtr)) ()
 
+> headTimeBuffer :: Ticks -> KernelR Bool
+> headTimeBuffer usage = do
+>     scPtr <- readCurSc
+>     sc <- readSchedContext scPtr
+>     return $ rAmount (refillHd sc) <= usage && (maxBound :: Word64) - rTime (refillHd sc) >= 5 * usToTicks maxPeriodUs
+
+> handleOverrunLoopBody :: Ticks -> Kernel Ticks
+> handleOverrunLoopBody usage = do
+>     scPtr <- getCurSc
+>     single <- refillSingle scPtr
+>     sc <- getSchedContext scPtr
+
+>     usage' <- return (usage - rAmount (refillHd sc))
+
+>     if single
+>        then updateRefillHd scPtr $ \r -> r { rTime = rTime r + scPeriod sc}
+>        else do
+>          old_head <- refillPopHead scPtr
+>          let new = old_head { rTime = rTime old_head + scPeriod sc}
+>          scheduleUsed scPtr new
+
+>     return usage'
+
+> handleOverrunLoop :: Ticks -> Kernel Ticks
+> handleOverrunLoop usage =
+>     whileLoop (\usage -> fromJust . runReaderT (headTimeBuffer usage)) (\usage -> handleOverrunLoopBody usage) usage
+
 > refillBudgetCheck :: Ticks -> Kernel ()
 > refillBudgetCheck usage = do
 >     scPtr <- getCurSc
+
 >     sc <- getSchedContext scPtr
 >     assert (scRefillMax sc > 0) "CurSc should be active"
 
->     usage' <- return (min usage (rAmount (refillHd sc)))
+>     usage' <- handleOverrunLoop usage
+
+>     sc <- getSchedContext scPtr
+>     newHeadAmount <- return (rAmount (refillHd sc))
 
->     when (usage' > 0) $ do
+>     when (usage' > 0 && usage' < newHeadAmount) $ do
```

Comment:
I requested Curtis change things around so that we always had `MIN_BUDGET` in the head. The changes are here

https://github.com/kent-mcleod/seL4/pull/10/files

He is also happy with `5` in the check, but I'm not sure that's been updated in the C anywhere. (The changes to `refill_update` aren't necessary any more, and I've let Curtis know about that, too)

Reply:
Ah, cool, that makes sense. I was looking at https://github.com/kent-mcleod/seL4/pull/7/files instead.

Reply:
Yeah sorry I should have done a better job listing the various C commits